lib/seattleflu/id3c/cli/command/etl/redcap_det_fh_airs.py

```diff
@@ -0,0 +1,857 @@
+            weekly_encounter_entry,
+            weekly_questionnaire_entry,
+            weekly_computed_questionnaire_entry
+        ]
+
+        persisted_resource_entries.extend(list(filter(None, current_instance_entries)))
+
+    return create_bundle_resource(
+        bundle_id = str(uuid4()),
+        timestamp = datetime.now().astimezone().isoformat(),
+        source = f"{REDCAP_URL}{enrollment.project.id}/{enrollment.id}",
+        entries = list(filter(None, persisted_resource_entries))
+    )
+
+
+def airs_get_encounter_date(record: REDCapRecord, event_type: EventType) -> Optional[str]:
+    encounter_date = None
+
+    if event_type == EventType.ENCOUNTER:
+        encounter_date = extract_date_from_survey_timestamp(record, 'weekly')
+    elif event_type == EventType.ENROLLMENT:
+        encounter_date = record.get('enr_date_complete') and \
+            datetime.strptime(record.get('enr_date_complete'), '%Y-%m-%d').strftime('%Y-%m-%d')
+    else:
+        # We should never get here, but we should also never have an 
+        #  if/elif without an else
+        LOG.error(f"Invalid date: {record.event_name!r} for record "
+                  f"{record.get('subject_id')} contains event_type {event_type}, "
+                  "but that is not one that we process.")
+    return encounter_date
+
+
+def create_encounter_id(record: REDCapRecord, is_followup_encounter: bool) -> str:
+    """
+    Create the encounter_id from the REDCap *record*.
+    """
+    if record.event_name:
+        redcap_event_name = record.event_name
+    else:
+        redcap_event_name = ''
+
+    if record.repeat_instance:
+        redcap_repeat_instance = str(record.repeat_instance)
+    else:
+        redcap_repeat_instance = ''
+
+    if is_followup_encounter:
+        encounter_identifier_suffix = "_follow_up"
+    else:
+        encounter_identifier_suffix = ''
+
+    return f'{record.project.base_url}{record.project.id}/{record.id}/{redcap_event_name}/' + \
+        f'{redcap_repeat_instance}{encounter_identifier_suffix}'
+
+
+def get_collection_date(record: REDCapRecord, collection_method: CollectionMethod) -> Optional[str]:
+    """
+    Determine sample/specimen collection date from the given REDCap *record*.
+    """
+    # For all surveys, try the survey _timestamp field (which is in Pacific time)
+    # before custom fields because the custom fields aren't always populated and when
+    # they are populated they use the browser's time zone.
+    collection_date = None
+
+    if collection_method == CollectionMethod.SWAB_AND_SEND:
+        collection_date = record.get("kit_reg_date_v2") \
+            or extract_date_from_survey_timestamp(record, "post_collection_data_entry_qc_2") \
+            or record.get("back_end_scan_v2") \
+            or record.get("kit_reg_date") \
+            or extract_date_from_survey_timestamp(record, "post_collection_data_entry_qc") \
+            or record.get("back_end_scan")
```

Comment:
A couple thoughts on this part.

* Should this be split out into two different functions or pathways depending if the swab is the 1 or 2 for the week? If two samples were collected in the same week, I think we'd need to be able to keep the collection dates straight and check the appropriate instrument.

* Can you comment on the instruments chosen for the sample collection timestamp and the ordering? I was thinking that for `swab_and_send` samples, we would be using the `airs_kit_activation` 1 and 2 respectively . It looks like this is also checking back end mail scans and post collection data entry qc forms -- are those intended as backups in case the participant didnt activate the kit?
I'm not sure if I have the whole picture here, but would it make the most sense to prioritize the kit activation instruments over the other dates, and additionally prioritize the survey instrument date before the custom fields within each instrument?